### PR TITLE
Ah gawb 1510: return groups for discoverability

### PR DESCRIPTION
**Pull request by ahaessly**
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [x] **LR** sign off
- [x] **Product Owner** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/LibraryApiServiceSpec.scala [hunk 1]
```diff
@@ -204,6 +205,15 @@ class LibraryApiServiceSpec extends BaseServiceSpec with LibraryApiService {
           }
         }
       }
+      "GET on " + libraryGroupsPath - {
+        "should return the all broad users group" in {
+          new RequestBuilder(HttpMethods.GET)(libraryGroupsPath) ~> dummyUserIdHeaders("1234") ~> sealRoute(libraryRoutes) ~> check {
+            status should equal(OK)
+            val respdata = response.entity.asString
+            assert(respdata.contains("all_broad_users"))
```
Comment (rushtong) [inline src/test/scala/org/broadinstitute/dsde/firecloud/service/LibraryApiServiceSpec.scala]: Instead of hard-coding this, it might be better to read the values from the config. This way, if we add more, the test doesn't have to change. Something like:
```
FireCloudConfig.ElasticSearch.discoverGroupNames.foreach { group => assert(respdata.contains(group)) }
```

**Review by rushtong — APPROVED** [review]
👍 LG, WG. One minor comment inline.

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/LibraryApiServiceSpec.scala (hunk 1 above)
Comment (davidangb) [reply src/test/scala/org/broadinstitute/dsde/firecloud/service/LibraryApiServiceSpec.scala]: even better, instead of checking `contains`, turn the expected and actual values into `Set`s, and check for equality between the sets.

File: src/main/resources/swagger/api-docs.yaml [hunk 2]
```diff
@@ -456,6 +456,17 @@ paths:
...
+    get:
+      tags:
+        - Library
+      operationId: groups
+      summary: |
+        returns groups that can be used for discoverable datasets
+      responses:
+        200:
+          description: Names of the allowed groups
+
```
Comment (davidangb) [inline src/main/resources/swagger/api-docs.yaml]: for clarity:
```
      responses:
        200:
          description: Names of the allowed groups
          schema:
            type: array
            items:
              type: string
```

**Review by davidangb — COMMENTED** [review]
couple minor comments inline, otherwise looking good!

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala [hunk 3]
```diff
@@ -54,14 +57,51 @@ class LibraryService (protected val argUserInfo: UserInfo, val rawlsDAO: RawlsDA
...
+    case PopulateSuggest(field: String, text: String) => populateSuggest(field: String, text: String) pipeTo sender
   }
 
+  def hasAccessOrCurator(workspaceResponse: WorkspaceResponse, neededLevel: WorkspaceAccessLevels.WorkspaceAccessLevel): Boolean = {
+    val wsaccess = workspaceResponse.accessLevel >= neededLevel
+    if (!wsaccess)
+      tryIsCurator(userInfo) map { value => return value }
+    wsaccess
+  }
+
```
Comment (ahaessly) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: this function doesn't work because i can't return the value from inside the map (i.e. the future).
i would like some sort of convenience function because we do these types of checks in several places below... any suggestions on what to do?

File: src/main/resources/swagger/api-docs.yaml [hunk 4]
```diff
@@ -533,6 +547,32 @@ paths:
       summary: Query to see if you're a library curator
       operationId: isCurator
 
+  /api/library/{workspaceNamespace}/{workspaceName}/discoverableBy:
```
Comment (rushtong) [inline src/main/resources/swagger/api-docs.yaml]: nitpick on endpoint name - I think it should be more noun-like for convention. `discoverableGroups` seems most appropriate since you're using that term already.

File: src/main/resources/swagger/api-docs.yaml [hunk 5]
```diff
@@ -533,6 +547,32 @@ paths:
...
+    put:
```
Comment (rushtong) [inline src/main/resources/swagger/api-docs.yaml]: Do we need a get? Not sure if that comes back as part of workspace or not.

Comment (rushtong) [reply src/main/resources/swagger/api-docs.yaml]: Also ... why isn't there a `POST` for this? Initially, one would be `POST`ing a new group and subsequent calls would be making a `PUT` request.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala [hunk 6]
```diff
@@ -54,14 +57,52 @@ class LibraryService (protected val argUserInfo: UserInfo, val rawlsDAO: RawlsDA
...
+    if (FireCloudConfig.ElasticSearch.discoverGroupNames.containsAll(newGroups.asJavaCollection)) {
+      rawlsDAO.getWorkspace(ns, name) map { workspaceResponse =>
+        if (workspaceResponse.accessLevel >= WorkspaceAccessLevels.Owner) {
+          // this is technically vulnerable to a race condition in which the workspace attributes have changed
+          // between the time we retrieved them and here, where we update them.
+          val remove = Seq(RemoveAttribute(discoverableWSAttribute))
+          val operations = newGroups map (group => AddListMember(discoverableWSAttribute, new AttributeString(group)))
+          RequestComplete(patchWorkspace(ns, name, remove ++ operations, isPublished(workspaceResponse)))
+        } else {
+          RequestCompleteWithErrorReport(Forbidden, "must be an owner to set discoverable By")
```
Comment (rushtong) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: nitpick: `Must be an owner to set the groups this workspace is discoverable by.` seems better. My inner grammar police is screaming not to end a sentence with a preposition, but apparently this is considered OK now.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala [hunk 7]
```diff
@@ -54,14 +57,52 @@ class LibraryService (protected val argUserInfo: UserInfo, val rawlsDAO: RawlsDA
...
+
+  def isPublished(workspaceResponse: WorkspaceResponse): Boolean = {
+    workspaceResponse.workspace.attributes.get(publishedFlag).fold(false)(_.asInstanceOf[AttributeBoolean].value)
+  }
+
+  /*
+   * This attribute is allowed to be modified by on owner or someone with canShare for the workspace
+   */
+  def updateDiscoverableBy(ns: String, name: String, newGroups: Seq[String]): Future[PerRequestMessage] = {
+    if (FireCloudConfig.ElasticSearch.discoverGroupNames.containsAll(newGroups.asJavaCollection)) {
```
Comment (rushtong) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: Could be rewritten without the java collection coercion:
```
if (newGroups.forall { g => FireCloudConfig.ElasticSearch.discoverGroupNames.contains(g) })
```

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala [hunk 8]
```diff
@@ -54,14 +57,52 @@ class LibraryService (protected val argUserInfo: UserInfo, val rawlsDAO: RawlsDA
...
+          // between the time we retrieved them and here, where we update them.
+          val remove = Seq(RemoveAttribute(discoverableWSAttribute))
+          val operations = newGroups map (group => AddListMember(discoverableWSAttribute, new AttributeString(group)))
+          RequestComplete(patchWorkspace(ns, name, remove ++ operations, isPublished(workspaceResponse)))
+        } else {
+          RequestCompleteWithErrorReport(Forbidden, "must be an owner to set discoverable By")
+        }
+      }
+    } else {
+      Future(RequestCompleteWithErrorReport(BadRequest, s"groups must be subset of allowable groups: %s".format(FireCloudConfig.ElasticSearch.discoverGroupNames)))
```
Comment (rushtong) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: When group names becomes more than one name, you'll want this to be better formatted:
```
FireCloudConfig.ElasticSearch.discoverGroupNames.toArray.mkString(", ")
```

**Review by rushtong — APPROVED** [review]
👍  LG, WG. Minor comments inline - feel free to address as you wish.

File: src/main/resources/swagger/api-docs.yaml (hunk 5 above)
Comment (ahaessly) [reply src/main/resources/swagger/api-docs.yaml]: this is really just setting a library value on an existing workspace. it's broken out into a separate call just for the API because it has different access control checks than the other library fields and these checks are handled when setting the value from the UI.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala [hunk 9]
```diff
@@ -54,14 +56,52 @@ class LibraryService (protected val argUserInfo: UserInfo, val rawlsDAO: RawlsDA
...
+      Future.successful(wsaccess)
+  }
+
+  def isPublished(workspaceResponse: WorkspaceResponse): Boolean = {
+    workspaceResponse.workspace.attributes.get(publishedFlag).fold(false)(_.asInstanceOf[AttributeBoolean].value)
+  }
+
+  /*
+   * This attribute is allowed to be modified by on owner or someone with canShare for the workspace
+   */
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: this comment doesn't match the implementation below.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala [hunk 10]
```diff
@@ -81,38 +121,65 @@ class LibraryService (protected val argUserInfo: UserInfo, val rawlsDAO: RawlsDA
...
-                k => k.namespace == AttributeName.libraryNamespace && k.name != LibraryService.publishedFlag.name)
-              val republish = workspaceResponse.workspace.attributes.get(publishedFlag).fold(false)(_.asInstanceOf[AttributeBoolean].value)
-              rawlsDAO.patchWorkspaceAttributes(ns, name, allOperations) map { newws =>
-                if (republish) {
-                  // we do not need to delete before republish
-                  publishDocument(newws)
-                }
-                RequestComplete(newws)
+            rawlsDAO.getWorkspace(ns, name) map  { workspaceResponse =>
+              if (workspaceResponse.accessLevel >= WorkspaceAccessLevels.Write) {
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: you shouldn't need this check; Rawls handles it. Is there a reason to enforce it here? Are you getting a head start on handling catalog permission?

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala [hunk 11]
```diff
@@ -81,38 +121,65 @@ class LibraryService (protected val argUserInfo: UserInfo, val rawlsDAO: RawlsDA
...
+            RequestComplete(ws)
+          }
+        } else {
+          // verify owner on workspace
+          if (workspaceResponse.accessLevel < WorkspaceAccessLevels.Write)
+            Future(RequestCompleteWithErrorReport(Forbidden, "must be a curator or have at least write privileges"))
           else
-            removeDocument(ws)
-          RequestComplete(ws)
+            Future(RequestCompleteWithErrorReport(Forbidden, "must be an owner or a curator"))
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: fyi - check out GAWB-1683. That ticket is *NOT* in scope for this PR - let's avoid scope creep and keep PRs small - mentioning it just so you're aware.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala [hunk 12]
```diff
@@ -81,38 +121,65 @@ class LibraryService (protected val argUserInfo: UserInfo, val rawlsDAO: RawlsDA
...
+            if (value)
+              publishDocument(ws)
+            else
+              removeDocument(ws)
+            RequestComplete(ws)
+          }
+        } else {
+          // verify owner on workspace
+          if (workspaceResponse.accessLevel < WorkspaceAccessLevels.Write)
+            Future(RequestCompleteWithErrorReport(Forbidden, "must be a curator or have at least write privileges"))
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: error message is now wrong re: curator - and in the case where the workspace isn't yet published, user would need owner privs.

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/LibraryApiServiceSpec.scala [hunk 13]
```diff
@@ -102,6 +105,13 @@ class LibraryApiServiceSpec extends BaseServiceSpec with LibraryApiService {
           }
         }
       }
+      "POST as writer on " + publishedPath() - {
+        "should be Forbidden" in {
+          new RequestBuilder(HttpMethods.POST)(publishedPath("unpublishedwriter")) ~> dummyUserIdHeaders("1234") ~> sealRoute(libraryRoutes) ~> check {
+            status should equal(Forbidden)
+          }
+        }
+      }
```
Comment (davidangb) [inline src/test/scala/org/broadinstitute/dsde/firecloud/service/LibraryApiServiceSpec.scala:114]: add a test for POST as writer on an already-published workspace, which should succeed.

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/LibraryApiServiceSpec.scala [hunk 14]
```diff
@@ -204,6 +224,15 @@ class LibraryApiServiceSpec extends BaseServiceSpec with LibraryApiService {
           }
         }
       }
+      "GET on " + libraryGroupsPath - {
+        "should return the all broad users group" in {
+          new RequestBuilder(HttpMethods.GET)(libraryGroupsPath) ~> dummyUserIdHeaders("1234") ~> sealRoute(libraryRoutes) ~> check {
+            status should equal(OK)
+            val respdata = response.entity.asString.parseJson.convertTo[Seq[String]]
+            assert(respdata.forall { g => FireCloudConfig.ElasticSearch.discoverGroupNames.contains(g) })
```
Comment (davidangb) [inline src/test/scala/org/broadinstitute/dsde/firecloud/service/LibraryApiServiceSpec.scala]: instead of testing `forall`, turn them into `Set`s and compare equality between the sets.

**Review by davidangb — COMMENTED** [review]
A few comments inline, I haven't tested!

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/LibraryApiServiceSpec.scala (hunk 14 above)
Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/firecloud/service/LibraryApiServiceSpec.scala]: The `forall` was my suggestion. I think what you need is to check that the set of `FireCloudConfig.ElasticSearch.discoverGroupNames` contains all of the elements in `respdata`. Simple set equality will fail as soon as we have more than one discoverable group and the user is only adding one of them to the workspace.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala (hunk 10 above)
Comment (ahaessly) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: i did have extra checks that rawls didn't do... but i'll fix since i'm not sure what the solution will be next

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala (hunk 11 above)
Comment (ahaessly) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: refactored this code. the removeDocument is for unpublishing not addressing 1683, but admit code was confusing. hopefully my changes are clearer, even if duplicating some code. let me know if you have a better way :)

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/LibraryApiServiceSpec.scala (hunk 14 above)
Comment (ahaessly) [reply src/test/scala/org/broadinstitute/dsde/firecloud/service/LibraryApiServiceSpec.scala]: greg, your response was for a different part of the code, and i made the change there. this is the test and these groups should always have the same members

Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/firecloud/service/LibraryApiServiceSpec.scala]: Thanks Andrea!

**Comment by ahaessly** [comment]
i don't think i force pushed, but i don't see my responses anymore. let me know if i haven't answered a question you would like me to

**Review by davidangb — APPROVED** [review]
👍 looks good!
